src/executor/nested_loop_join.rs

```diff
@@ -21,170 +21,94 @@ pub struct NestedLoopJoinExecutor {
-        right_types: Vec<DataType>,
-    ) -> Result<Option<DataChunk>, ExecutorError> {
-        let left_row_len = left_types.len();
-        let right_row_len = left_types.len();
-
-        let mut chunk_builders: Vec<ArrayBuilderImpl> = vec![];
-        for ty in &left_types {
-            chunk_builders.push(ArrayBuilderImpl::new(ty));
-        }
-        for ty in &right_types {
-            chunk_builders.push(ArrayBuilderImpl::new(ty));
-        }
-
-        let mut left_bitmaps: Option<Vec<BitVec>> = match &join_op {
-            BoundJoinOperator::LeftOuter | BoundJoinOperator::FullOuter => {
-                let mut vecs = vec![];
-                for left_chunk in &left_chunks {
-                    vecs.push(bitvec![0; left_chunk.cardinality()]);
-                }
-                Some(vecs)
-            }
-            _ => None,
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        // collect all chunks from children
+        let left_chunks: Vec<DataChunk> = self.left_child.try_collect().await?;
+        let right_chunks: Vec<DataChunk> = self.right_child.try_collect().await?;
+
+        // helper functions
+        let create_builders = || {
+            self.left_types
+                .iter()
+                .chain(self.right_types.iter())
+                .map(|ty| ArrayBuilderImpl::with_capacity(PROCESSING_WINDOW_SIZE, ty))
+                .collect_vec()
         };
-
-        let mut right_bitmaps: Option<Vec<BitVec>> = match &join_op {
-            BoundJoinOperator::RightOuter | BoundJoinOperator::FullOuter => {
-                let mut vecs = vec![];
-                for right_chunk in &right_chunks {
-                    vecs.push(bitvec![0; right_chunk.cardinality()]);
+        let left_rows = || left_chunks.iter().flat_map(|chunk| chunk.rows());
+        let right_rows = || right_chunks.iter().flat_map(|chunk| chunk.rows());
+
+        // cross join: left x right
+        let mut builders = create_builders();
+        for left_row in left_rows() {
+            for right_row in right_rows() {
+                let values = left_row.values().chain(right_row.values());
+                for (builder, v) in builders.iter_mut().zip(values) {
+                    builder.push(&v);
                 }
-                Some(vecs)
             }
-            _ => None,
+        }
+        let cross_chunk = builders.into_iter().collect();
+
+        // evaluate filter bitmap
+        let filter = match self.condition.eval_array(&cross_chunk)? {
+            ArrayImpl::Bool(a) => a,
+            _ => panic!("unsupported value from join condition"),
         };
+        yield cross_chunk.filter(filter.iter().map(|b| matches!(b, Some(true))));
 
-        for left_chunk_idx in 0..left_chunks.len() {
-            for left_row_idx in 0..left_chunks[left_chunk_idx].cardinality() {
+        // append rows for left outer join
+        if matches!(
+            self.join_op,
```

Comment:
I prefer treat join_op as const generics, but we can refine it later.

Reply:
Sounds good. But if we do that, we have to construct 4 types when building the executor...

Reply:
Yes, @yuhao-su has done it in another repo.

Reply:
We can do it after adt_const_generics completed.